tests/e2e/test_lifecycle.py

```
@@ -0,0 +1,159 @@
+    result = vm.run("exit 7")
+    assert result.exit_code == 7
+
+
+def test_upload_download(vm: SmolVM, tmp_path: Path) -> None:
+    """A file round-trips host -> guest -> host byte-for-byte."""
+    payload = os.urandom(4096)
+    src = tmp_path / "payload.bin"
+    src.write_bytes(payload)
+    expected = hashlib.sha256(payload).hexdigest()
+
+    vm.upload_file(str(src), "/tmp/payload.bin")
```

Ruff (0.15.15)

[error] Probable insecure usage of temporary file or directory: "/tmp/payload.bin"

(S108)